monai/apps/reconstruction/transforms/dictionary.py

```
@@ -60,7 +63,12 @@ def __call__(self, data: Mapping[Hashable, NdarrayOrTensor]) -> dict[Hashable, T
         d = dict(data)
         for key in self.keys:
             if key in d[self.meta_key]:
-                d[key] = d[self.meta_key][key]  # type: ignore
+                extracted_value = d[self.meta_key][key]
+                # When image_only is True, skip if the extracted value is a MetaTensor
+                # to preserve metadata associations
+                if self.image_only and isinstance(extracted_value, MetaTensor):
```

Ruff (0.15.1)

[error] Undefined name `MetaTensor`

(F821)